src/codegen.cpp

```diff
@@ -1225,17 +1225,19 @@ jl_generic_fptr_t jl_generate_fptr(jl_method_instance_t *li, void *_F, size_t wo
     return fptr;
 }
 
-static Function *jl_cfunction_object(jl_function_t *f, jl_value_t *rt, jl_tupletype_t *argt);
+static Function *jl_cfunction_object(jl_function_t *f, jl_value_t *rt, jl_tupletype_t *argt, size_t world);
 // get the address of a C-callable entry point for a function
 extern "C" JL_DLLEXPORT
 void *jl_function_ptr(jl_function_t *f, jl_value_t *rt, jl_value_t *argt)
 {
+    jl_ptls_t ptls = jl_get_ptls_states();
     JL_GC_PUSH1(&argt);
     if (jl_is_tuple(argt)) {
         // TODO: maybe deprecation warning, better checking
         argt = (jl_value_t*)jl_apply_tuple_type_v((jl_value_t**)jl_data_ptr(argt), jl_nfields(argt));
     }
-    Function *llvmf = jl_cfunction_object(f, rt, (jl_tupletype_t*)argt);
+    Function *llvmf = jl_cfunction_object(f, rt, (jl_tupletype_t*)argt,
+                                          ptls->world_age);
```

Comment:
I think it's very important to make sure that we can still do the same `cfunction` optimization that we are doing now, i.e. `cfunction` on known types/functions should return a constant (or relocated) pointer. How can you do it if `cfunction` always returns the function in the latest world?